# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [3]:
# Pandas library in Python to manipulate and analyze the data

import pandas as pd

In [4]:
# Load the dataset from the provided link using the Pandas library

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
data = pd.read_csv(url)
data

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [5]:
# Find out the number of rows and columns in the dataset using the .shape attribute

num_rows, num_columns = data.shape
print(f'Number of rows: {num_rows}')
print(f'Number of columns: {num_columns}')

Number of rows: 4008
Number of columns: 11


In [6]:
# Standarize column names

data.columns = data.columns.str.lower().str.replace(" ", "_")

data = data.rename(columns={"st": "state"}) 
data.head()

customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [7]:
# Determine the data types of each column using the .dtypes method

data_types = data.dtypes
print(data_types)

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object


- **Customer Lifetime Value:** this column is represented as an object, but it should be converted to a numeric data type (possibly float64) if it's used for numerical calculations, as it represents the lifetime value of the customer.

- **Number of Open Complaints:** this column is represented as an object. If this column contains numerical information about the number of open complaints, it should be converted to a numeric data type, such as int64.

In summary, most of the data types seem appropriate based on the column definitions, but it's important to correct the data type of the CLV and Number of Open Complaints that should be converted to a numeric data type instead of an object.

**Suggestions for fixing incorrect data types**

**1. Identify the Issue:**

- Start by identifying the data type issue. Is the column supposed to be numeric but is being treated as text?

**2. Examine the Column:**

- Use df['column_name'].head() to examine the first few rows of the column and check for unexpected or invalid values. This could cause data type issues.

**3. Identify Anomalous Values:**

- If you find anomalous values, such as text in a numeric column or incorrectly formatted dates, identify them and make a note of what they are.

**4. Correct Incorrect Values:**

- If you identified incorrect values, you can either correct them or remove them as appropriate. For example, you can use df.loc[] to select rows with incorrect values and then correct or remove them.

**5. Change Data Type:**

- Use pandas conversion functions like pd.to_numeric() or astype() to change the data type of the column. Ensure that the new data type is appropriate for the values the column should contain.

**6. Verify Results:**

- Verify that the conversion was successful using df['column_name'].dtype. It should display the new data type.

**7. Save Changes (Optional):**

- If you're confident that the changes are correct and you don't want to revert them, you can save the changes to the original DataFrame or to a new DataFrame as needed.

In [8]:
# Identify the number of unique values and categorical columns:
# Use df.nunique() to count the number of unique values in each column. 
# Consider columns with a low number of unique values as candidates for being categorical. 
# Inspect the unique categories in these columns.

In [9]:
# Count unique values in each column
unique_counts = data.nunique()
unique_counts

customer                     1071
state                           8
gender                          5
education                       6
customer_lifetime_value      1027
income                        774
monthly_premium_auto          132
number_of_open_complaints       6
policy_type                     3
vehicle_class                   6
total_claim_amount            761
dtype: int64

In [10]:
# Identify categorical columns (those with fewer than a certain number of unique values)
categorical_cols = unique_counts[unique_counts < 10].index
categorical_cols

Index(['state', 'gender', 'education', 'number_of_open_complaints',
       'policy_type', 'vehicle_class'],
      dtype='object')

In [11]:
# Display unique categories of categorical columns
for col in categorical_cols:
    print(f"Column '{col}': {data[col].unique()}")

Column 'state': ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
Column 'gender': [nan 'F' 'M' 'Femal' 'Male' 'female']
Column 'education': ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
Column 'number_of_open_complaints': ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]
Column 'policy_type': ['Personal Auto' 'Corporate Auto' 'Special Auto' nan]
Column 'vehicle_class': ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


In [12]:
# Determine the range of values for numerical columns computing the minimum and maximum values for each numerical column. 
# This will provide insights into the spread or dispersion of data within each numerical variable.

In [13]:
# Select numerical columns in a new variable 
numerical_columns = data.select_dtypes(include=['int64', 'float64'])  
numerical_columns

income  monthly_premium_auto  total_claim_amount
0         0.0                1000.0            2.704934
1         0.0                  94.0         1131.464935
2     48767.0                 108.0          566.472247
3         0.0                 106.0          529.881344
4     36357.0                  68.0           17.269323
...       ...                   ...                 ...
4003      NaN                   NaN                 NaN
4004      NaN                   NaN                 NaN
4005      NaN                   NaN                 NaN
4006      NaN                   NaN                 NaN
4007      NaN                   NaN                 NaN

[4008 rows x 3 columns]

In [14]:
# Compute the minimum and maximum values for each numerical column
min_values = numerical_columns.min()
max_values = numerical_columns.max()

In [15]:
# Calculate the range for each numerical column
value_range = max_values - min_values

# Display the range for each numerical column
print("Range of Values for Numerical Columns:")
print(value_range)

Range of Values for Numerical Columns:
income                  99960.000000
monthly_premium_auto    35293.000000
total_claim_amount       2892.857571
dtype: float64


**Insights:**

- For the **"income"** column, the range of values spans from 0 to 99960. This suggests a wide variation in customer incomes, ranging from lower incomes to relatively high incomes.

- The **"monthly_premium_auto"** column has a range of values from 0 to 35,293. This indicates that the monthly premium amounts vary significantly, with some customers paying low premiums and others paying relatively high premiums.

- In the case of the **"total_claim_amount"** column, the range extends from 0 to approximately 2,892.86. This indicates a wide variation in the total claim amounts made by customers, with some claims being small and others relatively larger.

In [16]:
# Summary statistics for numerical columns

numeric_summary = data.describe()
print(numeric_summary)

             income  monthly_premium_auto  total_claim_amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678


- **"income":** The data shows that the average income is approximately $ 39,296, with a wide range of values from $ 0 to $ 99,960. The standard deviation is relatively high, indicating significant income variability among customers.

- **"monthly_premium_auto":** The mean monthly premium is around $ 193, with a wide range of values from $ 61 to $ 35,354. The standard deviation is quite large, suggesting substantial variability in premium amounts.

- **"total_claim_amount":** On average, the total claim amount is approximately $ 405, with values ranging from $ 0.38 to $ 2,893.24. The standard deviation reflects variability in claim amounts.

In [17]:
# Summary statistics for categorical columns

categorical_summary = data.describe(include='object')
print(categorical_summary)

       customer   state gender education customer_lifetime_value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       number_of_open_complaints    policy_type  vehicle_class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576  


- **"customer":** Each customer has a unique identifier (customer ID). There are 1,071 unique customers in the dataset, with each occurring only once.

- **"state":** There are 8 unique states represented in the dataset, with Oregon being the most frequent state (occurs 320 times).

- **"gender":** There are 5 unique gender categories. The most frequent gender is Female (occurs 457 times).

- **"education":** There are 6 unique education levels, with Bachelor being the most frequent (occurs 324 times).

- **"customer_lifetime_value":** There are 1,027 unique values for customer lifetime value, but there are 4 instances where the value "445811.34%" occurs multiple times.

- **"number_of_open_complaints":** There are 6 unique complaint categories. The most frequent category is "1/0/00" (occurs 830 times).

- **"policy_type":** There are 3 unique policy types, with "Personal Auto" being the most frequent (occurs 780 times).

- **"vehicle_class":** There are 6 unique vehicle classes, with "Four-Door Car" being the most frequent (occurs 576 times).

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [18]:
import pandas as pd

In [19]:
# Create a Series object with customer locations and their frequencies
location_counts = data['state'].value_counts()
location_counts

state
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [20]:
# Sort the locations in ascending order of frequency
top_5_less_common_locations = location_counts.sort_values().head(5)
print(top_5_less_common_locations)

state
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [25]:
# Count the number of policies sold for each policy type
policy_counts = data['policy_type'].value_counts()
policy_counts

policy_type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [27]:
# Find the policy type with the highest number of policies sold
most_sold_policy = policy_counts.idxmax()
most_sold_policy

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [28]:
# Create a DataFrame for Personal Auto policies
personal_auto_df = data.loc[data['policy_type'] == 'Personal Auto']
personal_auto_df

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
4     GA49547  Washington      M  High School or Below   
5     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
1065  XD85577        Cali      M              Bachelor   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1070  MY31220  California      F               College   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
4                 536307.65%  36357.0                  68.0   
5                 825629.78%  62902.0                  69.0   
...                      ...      ...                   ...   
1065              284226.69%  69417.0                  73.0   
1066              305955.03%  38644.0                  78.0   
1067             2031499.76%  63209.0                 102.0   
1068              323912.47%  16061.0                  88.0   
1070              899704.02%  54230.0                 112.0   

     number_of_open_complaints    policy_type  vehicle_class  \
0                       1/0/00  Personal Auto  Four-Door Car   
1                       1/0/00  Personal Auto  Four-Door Car   
2                       1/0/00  Personal Auto   Two-Door Car   
4                       1/0/00  Personal Auto  Four-Door Car   
5                       1/0/00  Personal Auto   Two-Door Car   
...                        ...            ...            ...   
1065                    1/1/00  Personal Auto  Four-Door Car   
1066                    1/1/00  Personal Auto  Four-Door Car   
1067                    1/2/00  Personal Auto            SUV   
1068                    1/0/00  Personal Auto  Four-Door Car   
1070                    1/0/00  Personal Auto   Two-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
4              17.269323  
5             159.383042  
...                  ...  
1065           30.874869  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1070          537.600000  

[780 rows x 11 columns]

In [29]:
# Create a DataFrame for Corporate Auto policies
corporate_auto_df = data.loc[data['policy_type'] == 'Corporate Auto']
corporate_auto_df

customer       state gender             education  \
3     WW63253  California      M              Bachelor   
6     XZ87318      Oregon      F               College   
7     CF85061     Arizona      M                Master   
8     DY87989      Oregon      M              Bachelor   
12    DP39365  California    NaN                Master   
...       ...         ...    ...                   ...   
1045  WZ31900      Oregon      F              Bachelor   
1050  BJ53923     Arizona      M  High School or Below   
1052  FB23788      Oregon      M  High School or Below   
1056  PI47776      Oregon      F              Bachelor   
1059  YG44474      Oregon      M               College   

     customer_lifetime_value   income  monthly_premium_auto  \
3                 764586.18%      0.0                 106.0   
6                 538089.86%  55350.0                  67.0   
7                 721610.03%      0.0                 101.0   
8                2412750.40%  14072.0                  71.0   
12                879879.70%  77026.0                 110.0   
...                      ...      ...                   ...   
1045              864970.06%  94389.0                 107.0   
1050              260027.21%  51978.0                  66.0   
1052              882883.50%  86721.0                 111.0   
1056              253070.51%  89451.0                  63.0   
1059             1401472.13%  54193.0                 117.0   

     number_of_open_complaints     policy_type  vehicle_class  \
3                       1/0/00  Corporate Auto            SUV   
6                       1/0/00  Corporate Auto  Four-Door Car   
7                       1/0/00  Corporate Auto  Four-Door Car   
8                       1/0/00  Corporate Auto  Four-Door Car   
12                      1/2/00  Corporate Auto  Four-Door Car   
...                        ...             ...            ...   
1045                    1/0/00  Corporate Auto            SUV   
1050                    1/0/00  Corporate Auto  Four-Door Car   
1052                    1/0/00  Corporate Auto            SUV   
1056                    1/0/00  Corporate Auto  Four-Door Car   
1059                    1/0/00  Corporate Auto            SUV   

      total_claim_amount  
3             529.881344  
6             321.600000  
7             363.029680  
8             511.200000  
12            472.029737  
...                  ...  
1045           85.063708  
1050          144.782152  
1052          532.800000  
1056           61.769564  
1059          720.752945  

[234 rows x 11 columns]

In [30]:
# Calculate the average income for Personal Auto policies
average_income_personal_auto = personal_auto_df['income'].mean()
average_income_personal_auto

38180.69871794872

In [31]:
# Calculate the average income for Corporate Auto policies
average_income_corporate_auto = corporate_auto_df['income'].mean()
average_income_corporate_auto

41390.31196581197

In [32]:
print(f'Average Income for Personal Auto Policies: ${average_income_personal_auto:.2f}')
print(f'Average Income for Corporate Auto Policies: ${average_income_corporate_auto:.2f}')

Average Income for Personal Auto Policies: $38180.70
Average Income for Corporate Auto Policies: $41390.31
